### PR TITLE
add protocol fees fixin

### PR DESCRIPTION
**Pull request by smarx**
## Description

This change adds `FixinProtocolFees`, code that collects protocol fees using per-pool `FeeCollector` contracts. Once per epoch, a function on each `FeeCollector` is called to transfer the held ether (first converted to WETH) to the staking contract. Delaying this transfer and associated accounting yields significant gas savings.

## Testing instructions

`yarn build && yarn test` in `contracts/zero-ex`.

The basic happy path with WETH is tested, but all other tests are missing. Future commits will add more robust testing, including negative testing for failure modes.

## Types of changes

* New feature (non-breaking change which adds functionality)

## Checklist:

-   [x] Prefix PR title with `[WIP]` if necessary.
-   [x] Add tests to cover changes as needed.
-   [x] Update documentation as needed.
-   [x] Add new entries to the relevant CHANGELOG.jsons.

### CODE REVIEW

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+import "@0x/contracts-erc20/contracts/src/v06/IEtherTokenV06.sol";
+import "../external/FeeCollector.sol";
+import "../features/libs/LibTokenSpender.sol";
+
+/// @dev Helpers for collecting protocol fees.
+abstract contract FixinProtocolFees {
+    bytes32 immutable feeCollectorCodeHash;
+
+    constructor() internal {
+        feeCollectorCodeHash = keccak256(type(FeeCollector).creationCode);
```
Comment (hysz) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:31]: Is the `immutable` cheaper than a `constant` in this case bc it'll fill in the final `bytes32` value instead of the `keccak256(...)?

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    )
+        internal
+    {
+        require(
+            msg.value == amount || msg.value == 0,
+            "FixinProtocolFees/WRONG_ETHER_AMOUNT");
+
+        FeeCollector feeCollector = _getFeeCollector(poolId);
+
+        if (msg.value == 0) {
```
Comment (hysz) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:52]: I suppose we're making the assumption that if the taker is selling ETH then their protocol fee must also be paid in ETH. I think this is probably fine, wdyt @dorothy-zbornak? 

... I feel like any workaround could get pretty hairy bc we'd have to track which ETH is for the fill vs the protocol fee. 

If we are going to make this assumption we should document it somewhere in the code, perhaps even in the function description for `_collectProtocolFee`.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 3]
```diff
@@ -0,0 +1,116 @@
...
+    /// @param weth The WETH token contract.
+    function _collectProtocolFee(
+        bytes32 poolId,
+        uint256 amount,
+        IERC20TokenV06 weth
+    )
+        internal
+    {
+        require(
+            msg.value == amount || msg.value == 0,
```
Comment (hysz) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:47]: I think we'll have to do `>= amount` in case the taker is also selling ETH or paying the fee multiple times.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 4]
```diff
@@ -0,0 +1,116 @@
...
+        FeeCollector feeCollector = _getFeeCollector(poolId);
+
+        uint256 codeSize;
+        assembly {
+            codeSize := extcodesize(feeCollector)
+        }
+
+        if (codeSize == 0) {
+            // Create and initialize the contract if necessary.
+            new FeeCollector{salt: poolId}();
```
Comment (hysz) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:80]: When we include `salt` does it implicitly use `create2`?

**Review by hysz — COMMENTED** [review]
Solidity is looking very clean! A few small comments/questions.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol [hunk 5]
```diff
@@ -0,0 +1,116 @@
...
+    function _collectProtocolFee(
+        bytes32 poolId,
+        uint256 amount,
+        IERC20TokenV06 weth
+    )
+        internal
+    {
+        require(
+            msg.value == amount || msg.value == 0,
+            "FixinProtocolFees/WRONG_ETHER_AMOUNT");
```
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:48]: Hmm, I wonder if we'll ever allow filling a WETH order with ETH, causing this to break. Guess we can cross that bridge when we get there.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 1 above)
Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:31]: Yeah, my understanding is that a constant with the value `keccak256(...)` will be reevaluated at each access. (I wish Solidity just disallowed it.) But the `immutable` can be computed just once.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 4 above)
Comment (dorothy-zbornak) [inline contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:80]: Syntax god.

**Review by dorothy-zbornak — APPROVED** [review]
wtg! I'll fix the changelog thing

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 2 above)
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:52]: I think it'll be fine for what we have roadmapped. Might need to tweak in the future.

Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:52]: Ah, I was borrowing the logic from https://github.com/0xProject/0x-monorepo/blob/development/contracts/staking/contracts/src/fees/MixinExchangeFees.sol#L155-L174. But I think I've overdone it... if ETH can be traded directly, my check that `msg.value == amount` won't work.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 3 above)
Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:47]: Ah, yes, great minds think alike. I just mentioned this in my response to the previous comment. :-) Yeah, this needs to be rethought.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 4 above)
Comment (smarx) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:80]: Yes, exactly.

File: contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol (hunk 3 above)
Comment (dorothy-zbornak) [reply contracts/zero-ex/contracts/src/fixins/FixinProtocolFees.sol:47]: I *think* there will be no case in our current roadmap where `msg.value != fee`, but I can change it if it does come up.

**Review by hysz — APPROVED** [review]
Nice work! Let's merge (\~‾▿‾)~
